File: R/trex_selector_methods/trex_gvs/demo_trex_gvs_07/demo_trex_gvs_07.R
# =================================================================================
# demo_trex_gvs_07.R
# =================================================================================
#
# T-Rex+GVS on the ARMA mixed-structure blocks DGP.
#
# Four blocks (sizes {20, 50, 80, 65}) each following a different time-series
# model:
#   Block 1 = AR(1)      [size 20]  — active
#   Block 2 = MA(3)      [size 50]  — active
#   Block 3 = ARMA(2,1)  [size 80]  — active   (s = 150)
#   Block 4 = AR(1) trap [size 65]  — inactive
# Shuffled into p = 500 columns with white-noise gaps. The AR component is
# governed by ar_coef; ma_coefs = c(0.5, 0.3, 0.1) is fixed. This is the most
# heterogeneous within-block structure in the suite.
#
# Methods compared: EN (TENET), EN+AUG (TENET_AUG), IEN.
# The swept correlation knob here is ar_coef (not rho).
#
# Parts:
#   Part 1: Single-run demo — MA(3) lag-structure check + block layout + one run.
#   Part 2: SNR sweep       — fixed ar_coef = 0.8.
#   Part 3: ar_coef sweep   — fixed SNR = 2.0.
#   Part 4: 2-D SNR x ar_coef grid.
#
# =================================================================================

library(TRexSelectorNeo)
library(plotly)
library(parallel)

num_cores <- local({
  a <- suppressWarnings(as.integer(commandArgs(trailingOnly = TRUE)[1L]))
  if (!is.na(a) && a > 0L) a else 6L
})

this_dir_ <- tryCatch(
  dirname(normalizePath(sys.frame(1)$ofile)),
  error = function(e) {
    args <- commandArgs(trailingOnly = FALSE)
    flag <- grep("^--file=", args, value = TRUE)
    if (length(flag) > 0L)
      dirname(normalizePath(sub("^--file=", "", flag[1L])))
    else "."
  }
)

source(file.path(this_dir_, "..", "..", "support_generators.R"))
source(file.path(this_dir_, "..", "..", "simulation_utils.R"))
source(file.path(this_dir_, "..", "trex_gvs_dgps.R"))


# ==============================================================================
# Configuration
# ==============================================================================

CFG <- list(
  n        = 200L,
  p        = 500L,
  ar_coef  = 0.8,          # fixed AR parameter for the SNR sweep
  snr      = 2.0,          # fixed SNR for the ar_coef sweep
  K        = 20L,
  tFDR     = 0.1,
  seed     = 2026L,
  num_MC   = 200L,
  corr_max = 0.98,
  hc_dist  = "single"
)

GVS_METHODS <- list(
  list(label = "EN",     GVS_type = "EN",  en_solver = "TENET"),
  list(label = "EN+AUG", GVS_type = "EN",  en_solver = "TENET_AUG"),
  list(label = "IEN",    GVS_type = "IEN", en_solver = "TENET")
)

run_part_1 <- TRUE
run_part_2 <- TRUE
run_part_3 <- TRUE
run_part_4 <- TRUE


# ==============================================================================
# Local helpers
# ==============================================================================

#' Print a formatted single-run result block.
.print_result <- function(scenario_name, dat, result, tFDR) {
  ts      <- which(dat$beta != 0)
  sel_set <- result$selected_indices
  n_sel   <- length(sel_set)
  n_tp    <- length(intersect(sel_set, ts))
  n_fp    <- n_sel - n_tp

  cat(strrep("=", 70), "\n")
  cat(sprintf("  %s\n", scenario_name))
  cat(strrep("-", 70), "\n")
  cat(sprintf(" Data: n = %d, p = %d, tFDR = %.2f, s = %d\n",
              dat$n, dat$p, tFDR, dat$s))
  cat(sprintf("       SNR = %.2f,  sigma_y = %.4f\n", dat$snr, dat$sigma_y))
  cat(strrep("-", 70), "\n")
  cat(sprintf("  Calibration:  T_stop = %d,  L = %d\n",
              result$T_stop, result$L))
  cat(sprintf("  Selection:    %d selected  |  TP = %d  FP = %d\n",
              n_sel, n_tp, n_fp))
  cat(sprintf("  Rates:        TPP = %.3f  |  FDP = %.3f  (target tFDR <= %.2f)\n",
              TRexSelectorNeo::compute_tpp(sel_set, ts),
              TRexSelectorNeo::compute_fdp(sel_set, ts), tFDR))
  cat(strrep("=", 70), "\n\n")
  invisible(NULL)
}

#' Run one MC point on the ARMA-blocks DGP for a given (method, snr, ar_coef).
.arma_point <- function(method, snr, ar_coef) {
  .run_mc_arma_blocks(
    n = CFG$n, p = CFG$p, snr = snr, ar_coef = ar_coef,
    tFDR = CFG$tFDR, K = CFG$K, num_MC = CFG$num_MC, seed = CFG$seed,
    GVS_type = method$GVS_type, corr_max = CFG$corr_max, hc_dist = CFG$hc_dist,
    en_solver = method$en_solver, n_cores = num_cores)
}


# ==============================================================================
# Part 1: Single-run demo
# ==============================================================================

if (run_part_1) {
  cat("\n", strrep("=", 70), "\n", sep = "")
  cat("ARMA-Blocks GVS  |  Part 1: Single-run\n")
  cat(sprintf("n=%d,  p=%d,  s=150  (blocks 20+50+80 active; 65 trap)\n",
              CFG$n, CFG$p))
  cat(sprintf("ar_coef=%.2f,  SNR=%.2f,  tFDR=%.2f\n",
              CFG$ar_coef, CFG$snr, CFG$tFDR))
  cat("Block types: AR(1) | MA(3) | ARMA(2,1) | AR(1) trap\n")
  cat(strrep("=", 70), "\n\n")

  cat("[Part 1] Generating ARMA-blocks data ...\n")
  dat_p1 <- dgp_arma_blocks_snr(n = CFG$n, p = CFG$p, snr = CFG$snr,
                                ar_coef = CFG$ar_coef, seed = CFG$seed)
  cat(sprintf("[Part 1] Active: %d  |  sigma_y = %.4f\n\n",
              dat_p1$s, dat_p1$sigma_y))

  cat("[Part 1] Block layout (shuffled order):\n")
  block_sizes  <- c(20L, 50L, 80L, 65L)
  block_types  <- c("AR(1)", "MA(3)", "ARMA(2,1)", "AR(1) trap")
  active_label <- c("active", "active", "active", "trap")
  for (b in seq_len(4L)) {
    idx <- dat_p1$block_indices[[b]]
    cat(sprintf("  Block %d [size=%2d, %s, %s]: columns %d - %d\n",
                b, block_sizes[b], block_types[b], active_label[b],
                min(idx), max(idx)))
  }
  cat(sprintf("  Block order (ID sequence placed): {%s}\n\n",
              paste(dat_p1$block_order, collapse = ", ")))

  # MA(3) lag structure: positive short-lag correlation, ~zero beyond lag 3.
  ma3_idx <- dat_p1$block_indices[[2L]]
  lag1 <- cor(dat_p1$X[, ma3_idx[1L]], dat_p1$X[, ma3_idx[2L]])
  lag4 <- cor(dat_p1$X[, ma3_idx[1L]], dat_p1$X[, ma3_idx[5L]])
  cat(sprintf("[Part 1] MA(3) Block 2 — Lag 1 correlation (expect > 0): %.4f\n", lag1))
  cat(sprintf("[Part 1] MA(3) Block 2 — Lag 4 (expect ~0):              %.4f\n\n", lag4))

  cormat_p1 <- plot_cormat(cor(dat_p1$X))
  if (interactive()) {
    cat("[Part 1] Displaying correlation matrix ...\n")
    print(cormat_p1)
  }

  for (m in GVS_METHODS) {
    cat(sprintf("[Part 1] Running trex+GVS (%s) ...\n\n", m$label))
    res <- .gvs_select(dat_p1$X, dat_p1$y, CFG$tFDR, CFG$K,
                       m$GVS_type, CFG$corr_max, CFG$hc_dist, m$en_solver)
    .print_result(sprintf("Part 1 — ARMA-Blocks GVS  [%s]", m$label),
                  dat_p1, res, CFG$tFDR)
  }
}


# ==============================================================================
# Part 2: MC simulation - SNR sweep
# ==============================================================================
# Fixed ar_coef = 0.8. Compares EN / EN+AUG / IEN.

if (run_part_2) {
  snr_grid <- c(0.1, 0.2, 0.5, 1.0, 2.0, 5.0)

  cat(strrep("=", 70), "\n")
  cat("ARMA-Blocks GVS MC — Part 2: SNR sweep\n")
  cat(sprintf("tFDR=%.2f  ar_coef=%.2f  n=%d  p=%d  %d MC\n",
              CFG$tFDR, CFG$ar_coef, CFG$n, CFG$p, CFG$num_MC))
  cat(strrep("=", 70), "\n\n")

  for (m in GVS_METHODS) {
    cat(sprintf("\n  [%s] SNR sweep, %d MC per point ...\n", m$label, CFG$num_MC))
    res <- lapply(snr_grid, function(s) {
      cat(sprintf("    SNR = %.2f ...\n", s))
      c(list(SNR = s), .arma_point(m, s, CFG$ar_coef))
    })
    cat(sprintf("\n  Results — %s:\n", m$label))
    .print_table(res, "SNR")
  }
}


# ==============================================================================
# Part 3: MC simulation - ar_coef sweep
# ==============================================================================
# Fixed SNR = 2.0. ar_coef passed directly to the DGP.

if (run_part_3) {
  ar_coef_grid <- c(0.10, 0.20, 0.30, 0.40, 0.50, 0.60, 0.70, 0.80, 0.90, 0.95)

  cat(strrep("=", 70), "\n")
  cat("ARMA-Blocks GVS MC — Part 3: ar_coef sweep\n")
  cat(sprintf("tFDR=%.2f  SNR=%.2f  n=%d  p=%d  %d MC\n",
              CFG$tFDR, CFG$snr, CFG$n, CFG$p, CFG$num_MC))
  cat(strrep("=", 70), "\n\n")

  for (m in GVS_METHODS) {
    cat(sprintf("\n  [%s] ar_coef sweep, %d MC per point ...\n",
                m$label, CFG$num_MC))
    res <- lapply(ar_coef_grid, function(ar_val) {
      cat(sprintf("    ar_coef = %.2f ...\n", ar_val))
      c(list(ar_coef = ar_val), .arma_point(m, CFG$snr, ar_val))
    })
    cat(sprintf("\n  Results — %s:\n", m$label))
    .print_table(res, "ar_coef", "%-8.2f")
  }
}


# ==============================================================================
# Part 4: 2-D phase-transition sweep - SNR x ar_coef
# ==============================================================================

if (run_part_4) {
  snr_grid     <- c(0.2, 0.5, 1.0, 2.0, 5.0)
  ar_coef_grid <- c(0.30, 0.50, 0.70, 0.80, 0.90, 0.95)
  row_nms      <- paste0("snr=", sprintf("%.2f", snr_grid))
  col_nms      <- paste0("ar=", sprintf("%.2f", ar_coef_grid))

  cat(strrep("=", 70), "\n")
  cat("ARMA-Blocks GVS MC  |  Part 4: SNR x ar_coef grid\n")
  cat(sprintf("n=%d  p=%d  %d MC  tFDR=%.2f  corr_max=%.2f\n",
              CFG$n, CFG$p, CFG$num_MC, CFG$tFDR, CFG$corr_max))
  cat(strrep("=", 70), "\n\n")

  for (m in GVS_METHODS) {
    mat_TPP <- matrix(NA_real_, length(snr_grid), length(ar_coef_grid),
                      dimnames = list(row_nms, col_nms))
    mat_FDP <- mat_TPP
    for (i in seq_along(snr_grid)) {
      for (j in seq_along(ar_coef_grid)) {
        cat(sprintf("  [%s] SNR=%.2f ar_coef=%.2f ...\n",
                    m$label, snr_grid[i], ar_coef_grid[j]))
        r <- .arma_point(m, snr_grid[i], ar_coef_grid[j])
        mat_TPP[i, j] <- r$mean_TPP
        mat_FDP[i, j] <- r$mean_FDP
      }
    }
    .print_matrix(mat_TPP,
                  sprintf("mean_TPP [%s] (rows: SNR, cols: ar_coef)", m$label))
    .print_matrix(mat_FDP,
                  sprintf("mean_FDP [%s] (rows: SNR, cols: ar_coef)", m$label))
  }
}


cat("\nARMA-blocks GVS demo complete.\n")

File: Python/trex_selector_methods/trex_gvs/demo_trex_gvs_07/demo_trex_gvs_07.py
# ==============================================================================
# demo_trex_gvs_07.py
# ==============================================================================
#
# T-Rex+GVS on the ARMA mixed-structure blocks DGP.
#
# Four blocks (sizes {20, 50, 80, 65}) each following a different time-series
# model:
#   Block 0 = AR(1)      [size 20]  — active
#   Block 1 = MA(3)      [size 50]  — active
#   Block 2 = ARMA(2,1)  [size 80]  — active   (s = 150)
#   Block 3 = AR(1) trap [size 65]  — inactive
# Shuffled into p = 500 columns with white-noise gaps. The AR component is
# governed by ar_coef; ma_coefs = (0.5, 0.3, 0.1) is fixed. This is the most
# heterogeneous within-block structure in the suite.
#
# Methods compared: EN (TENET), EN+AUG (TENET_AUG), IEN.
# The swept correlation knob here is ar_coef (not rho).
#
# Parts:
#   Part 1: Single-run demo — MA(3) lag-structure check + block layout + one run.
#   Part 2: SNR sweep       — fixed ar_coef = 0.8.
#   Part 3: ar_coef sweep   — fixed SNR = 2.0.
#   Part 4: 2-D SNR x ar_coef grid.
#
# Mirrors R/trex_selector_methods/trex_gvs/demo_trex_gvs_07. Part 1 runs by
# default; the heavy MC parts are opt-in via RUN_PART_*.
# Indices are 0-based (Python convention); the R counterpart is 1-based.
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

from trex_gvs_dgps import dgp_arma_blocks_snr
from gvs_sim_common import (N_CORES, build_gvs_selector, run_mc_gvs,
                            print_table, print_matrix, print_single_run_result)

N_WORKERS = int(sys.argv[1]) if len(sys.argv) > 1 and sys.argv[1].isdigit() else N_CORES

CFG = dict(
    n=200, p=500,
    ar_coef=0.8,            # fixed AR parameter for the SNR sweep
    snr=2.0,                # fixed SNR for the ar_coef sweep
    K=20, tFDR=0.1,
    seed=2026, num_MC=200,
    corr_max=0.98, hc_dist="single",
)

GVS_METHODS = [
    dict(label="EN",     gvs_type="EN",  en_solver="TENET"),
    dict(label="EN+AUG", gvs_type="EN",  en_solver="TENET_AUG"),
    dict(label="IEN",    gvs_type="IEN", en_solver="TENET"),
]

RUN_PART_1 = True
RUN_PART_2 = False
RUN_PART_3 = False
RUN_PART_4 = False


def _arma_point(method, snr, ar_coef):
    """Run one MC point on the ARMA-blocks DGP for (method, snr, ar_coef)."""
    return run_mc_gvs(
        "dgp_arma_blocks_snr",
        dict(n=CFG["n"], p=CFG["p"], snr=snr, ar_coef=ar_coef),
        CFG["tFDR"], CFG["K"], CFG["num_MC"], CFG["seed"],
        method["gvs_type"], CFG["corr_max"], hc_linkage=CFG["hc_dist"],
        en_solver=method["en_solver"], n_cores=N_WORKERS)


# ==============================================================================
# Part 1: Single-run demo
# ==============================================================================

def part_1_single_run():
    print("\n" + "=" * 70)
    print("ARMA-Blocks GVS  |  Part 1: Single-run")
    print(f"n={CFG['n']},  p={CFG['p']},  s=150  (blocks 20+50+80 active; 65 trap)")
    print(f"ar_coef={CFG['ar_coef']:.2f},  SNR={CFG['snr']:.2f},  tFDR={CFG['tFDR']:.2f}")
    print("Block types: AR(1) | MA(3) | ARMA(2,1) | AR(1) trap")
    print("=" * 70 + "\n")

    print("[Part 1] Generating ARMA-blocks data ...")
    dat = dgp_arma_blocks_snr(n=CFG["n"], p=CFG["p"], snr=CFG["snr"],
                              ar_coef=CFG["ar_coef"], seed=CFG["seed"])
    print(f"[Part 1] Active: {dat['s']}  |  sigma_y = {dat['sigma_y']:.4f}\n")

    print("[Part 1] Block layout (shuffled order):")
    block_sizes = [20, 50, 80, 65]
    block_types = ["AR(1)", "MA(3)", "ARMA(2,1)", "AR(1) trap"]
    active_label = ["active", "active", "active", "trap"]
    for b in range(4):
        idx = dat["block_indices"][b]
        print(f"  Block {b} [size={block_sizes[b]:2d}, {block_types[b]}, {active_label[b]}]:"
              f" columns {int(idx.min())} - {int(idx.max())}")
    print(f"  Block order (ID sequence placed): {{{', '.join(map(str, dat['block_order']))}}}\n")

    # MA(3) lag structure: positive short-lag correlation, ~zero beyond lag 3.
    ma3 = dat["block_indices"][1]
    X = dat["X"]
    lag1 = np.corrcoef(X[:, ma3[0]], X[:, ma3[1]])[0, 1]
    lag4 = np.corrcoef(X[:, ma3[0]], X[:, ma3[4]])[0, 1]
    print(f"[Part 1] MA(3) Block 1 - Lag 1 correlation (expect > 0): {lag1:.4f}")
    print(f"[Part 1] MA(3) Block 1 - Lag 4 (expect ~0):              {lag4:.4f}\n")

    for m in GVS_METHODS:
        print(f"[Part 1] Running trex+GVS ({m['label']}) ...\n")
        sel = build_gvs_selector(dat["X"], dat["y"], CFG["tFDR"], CFG["K"],
                                 m["gvs_type"], CFG["corr_max"], CFG["hc_dist"],
                                 m["en_solver"])
        print_single_run_result(f"Part 1 - ARMA-Blocks GVS  [{m['label']}]",
                                 dat, sel, CFG["tFDR"])


# ==============================================================================
# Part 2: MC simulation - SNR sweep  (fixed ar_coef = 0.8)
# ==============================================================================

def part_2_snr_sweep():
    snr_grid = [0.1, 0.2, 0.5, 1.0, 2.0, 5.0]

    print("=" * 70)
    print("ARMA-Blocks GVS MC - Part 2: SNR sweep")
    print(f"tFDR={CFG['tFDR']:.2f}  ar_coef={CFG['ar_coef']:.2f}  n={CFG['n']}"
          f"  p={CFG['p']}  {CFG['num_MC']} MC")
    print("=" * 70 + "\n")

    for m in GVS_METHODS:
        print(f"\n  [{m['label']}] SNR sweep, {CFG['num_MC']} MC per point ...")
        rows = []
        for s in snr_grid:
            print(f"    SNR = {s:.2f} ...")
            rows.append({"SNR": s, **_arma_point(m, s, CFG["ar_coef"])})
        print(f"\n  Results - {m['label']}:")
        print_table(rows, "SNR")


# ==============================================================================
# Part 3: MC simulation - ar_coef sweep  (fixed SNR = 2.0)
# ==============================================================================

def part_3_ar_coef_sweep():
    ar_coef_grid = [0.10, 0.20, 0.30, 0.40, 0.50, 0.60, 0.70, 0.80, 0.90, 0.95]

    print("=" * 70)
    print("ARMA-Blocks GVS MC - Part 3: ar_coef sweep")
    print(f"tFDR={CFG['tFDR']:.2f}  SNR={CFG['snr']:.2f}  n={CFG['n']}"
          f"  p={CFG['p']}  {CFG['num_MC']} MC")
    print("=" * 70 + "\n")

    for m in GVS_METHODS:
        print(f"\n  [{m['label']}] ar_coef sweep, {CFG['num_MC']} MC per point ...")
        rows = []
        for ar in ar_coef_grid:
            print(f"    ar_coef = {ar:.2f} ...")
            rows.append({"ar_coef": ar, **_arma_point(m, CFG["snr"], ar)})
        print(f"\n  Results - {m['label']}:")
        print_table(rows, "ar_coef", value_fmt="{:<8.2f}", width=8)


# ==============================================================================
# Part 4: 2-D phase-transition sweep - SNR x ar_coef
# ==============================================================================

def part_4_snr_ar_grid():
    snr_grid = [0.2, 0.5, 1.0, 2.0, 5.0]
    ar_coef_grid = [0.30, 0.50, 0.70, 0.80, 0.90, 0.95]
    row_labels = [f"snr={s:.2f}" for s in snr_grid]
    col_labels = [f"ar={a:.2f}" for a in ar_coef_grid]

    print("=" * 70)
    print("ARMA-Blocks GVS MC  |  Part 4: SNR x ar_coef grid")
    print(f"n={CFG['n']}  p={CFG['p']}  {CFG['num_MC']} MC  tFDR={CFG['tFDR']:.2f}"
          f"  corr_max={CFG['corr_max']:.2f}")
    print("=" * 70 + "\n")

    for m in GVS_METHODS:
        mat_tpp = np.full((len(snr_grid), len(ar_coef_grid)), np.nan)
        mat_fdp = np.full_like(mat_tpp, np.nan)
        for i, s in enumerate(snr_grid):
            for j, ar in enumerate(ar_coef_grid):
                print(f"  [{m['label']}] SNR={s:.2f} ar_coef={ar:.2f} ...")
                r = _arma_point(m, s, ar)
                mat_tpp[i, j] = r["mean_TPP"]
                mat_fdp[i, j] = r["mean_FDP"]
        print_matrix(mat_tpp, f"mean_TPP [{m['label']}] (rows: SNR, cols: ar_coef)",
                     row_labels, col_labels)
        print_matrix(mat_fdp, f"mean_FDP [{m['label']}] (rows: SNR, cols: ar_coef)",
                     row_labels, col_labels)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    if RUN_PART_1:
        part_1_single_run()
    if RUN_PART_2:
        part_2_snr_sweep()
    if RUN_PART_3:
        part_3_ar_coef_sweep()
    if RUN_PART_4:
        part_4_snr_ar_grid()
    print("\nARMA-blocks GVS demo complete.")
